File: src/net/reichholf/dreamdroid/parsers/enigma2/saxhandler/E2VolumeHandler.java
/* © 2010 Stephan Reichholf <stephan at reichholf dot net>
 * 
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */

package net.reichholf.dreamdroid.parsers.enigma2.saxhandler;

import net.reichholf.dreamdroid.helpers.ExtendedHashMap;
import net.reichholf.dreamdroid.helpers.enigma2.Volume;

import org.xml.sax.Attributes;
import org.xml.sax.helpers.DefaultHandler;

public class E2VolumeHandler extends DefaultHandler{
	private boolean inResult;
	private boolean inCurrent;
	private boolean inMuted;

	private ExtendedHashMap mResult;

	/**
	 * @param list
	 */
	public E2VolumeHandler(ExtendedHashMap res) {
		mResult = res;
	}

	/* (non-Javadoc)
	 * @see org.xml.sax.helpers.DefaultHandler#startElement(java.lang.String, java.lang.String, java.lang.String, org.xml.sax.Attributes)
	 */
	@Override
	public void startElement(String namespaceUri, String localName, String qName, Attributes attrs) {
		if (localName.equals("e2result")) {
			inResult = true;
		} else if (localName.equals("e2current")) {
			inCurrent = true;
		} else if (localName.equals("e2ismuted")) {
			inMuted = true;
		}
	}

	/* (non-Javadoc)
	 * @see org.xml.sax.helpers.DefaultHandler#endElement(java.lang.String, java.lang.String, java.lang.String)
	 */
	@Override
	public void endElement(String namespaceURI, String localName, String qName) {
		if (localName.equals("e2result")) {
			inResult = false;
		} else if (localName.equals("e2current")) {
			inCurrent = false;
		} else if (localName.equals("e2ismuted")) {
			inMuted = false;
		}
	}

	/* (non-Javadoc)
	 * @see org.xml.sax.helpers.DefaultHandler#characters(char[], int, int)
	 */
	@Override
	public void characters(char ch[], int start, int length) {
		String value = new String(ch, start, length);

		if (inResult) {
			mResult.putOrConcat(Volume.RESULT, value);
		} else if(inCurrent){
			mResult.putOrConcat(Volume.CURRENT, value);
		} else if(inMuted){
			mResult.putOrConcat(Volume.MUTED, value);
		}
	}
}

File: app/src/net/reichholf/dreamdroid/parsers/enigma2/saxhandler/E2VolumeHandler.kt
/* © 2010 Stephan Reichholf <stephan at reichholf dot net>
 *
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */

package net.reichholf.dreamdroid.parsers.enigma2.saxhandler

import net.reichholf.dreamdroid.helpers.enigma2.Volume
import org.xml.sax.Attributes

class E2VolumeHandler : E2SimpleHandler() {
    private var inResult: Boolean = false
    private var inCurrent: Boolean = false
    private var inMuted: Boolean = false

    override fun startElement(
        namespaceUri: String?,
        localName: String,
        qName: String?,
        attrs: Attributes?,
    ) {
        when (localName) {
            TAG_E2RESULT -> inResult = true
            TAG_E2CURRENT -> inCurrent = true
            TAG_E2ISMUTED -> inMuted = true
        }
    }

    override fun endElement(namespaceURI: String?, localName: String, qName: String?) {
        when (localName) {
            TAG_E2RESULT -> inResult = false
            TAG_E2CURRENT -> inCurrent = false
            TAG_E2ISMUTED -> inMuted = false
        }
    }

    override fun characters(ch: CharArray, start: Int, length: Int) {
        val value = String(ch, start, length)
        when {
            inResult -> mResult!!.putOrConcat(Volume.KEY_RESULT, value)
            inCurrent -> mResult!!.putOrConcat(Volume.KEY_CURRENT, value)
            inMuted -> mResult!!.putOrConcat(Volume.KEY_MUTED, value)
        }
    }

    companion object {
        @JvmStatic
        protected val TAG_E2ISMUTED: String = "e2ismuted"
        @JvmStatic
        protected val TAG_E2CURRENT: String = "e2current"
        @JvmStatic
        protected val TAG_E2RESULT: String = "e2result"
    }
}
